### PR TITLE
Add @SafeVarargs to ExternalPluginManager.loadBuiltin

### PR DESCRIPTION
**Pull request by LootBagger**
Fixes warning "Unchecked generics array creation for varargs parameter" on
loadBuiltin. This warning is trying to prevent a RuntimeException which
happens iff the varargs are modified in a particular way. Varargs are not
modified, which is guaranteed by PluginManager.loadPlugins using a Stream.

Change loadPlugins to take a Stream of Plugins. Saves a few allocations.

Allow loading of plugins that are not direct subclasses of Plugin, but
subclasses of subclasses of Plugin, etc. (fix in loadPlugins) and add a
test case that this works

Fix an iterator invalidation bug in PluginManager.topologicalSort.
Graph.sucessors is an iterator; removing nodes while iterating throws
a concurrentModificationException. Fix: add an allocation.

Fill out existing javadocs for topologicalSort

### CODE REVIEW

**Comment by LootBagger** [comment]
This is my first PR here. I originally imagined this PR as being simple, but went down the rabbit hole on something the Runelite devs probably don't care much about. I won't be upset if it gets closed, but I'd appreciate some explanation of what I did right/wrong here.

**Comment by Adam-** [comment]
I don't think I see a good reason to allow Plugin subclassing, it seems like it has no advantage but just opens a can of worms? And also a construct like
```java
abstract class BasePlugin extends Plugin{}
@PluginDescriptor(...)
class ChildPlugin extends BasePlugin{}
```
would trip the "Class is a plugin but has no plugin descriptor" warning.

**Comment by LootBagger** [comment]
Hm... I can't think of a good reason to do what you wrote in that example. Maybe a plugin hub plugin could extend a Runelite plugin to add some features to it. Or one author could create two versions of essentially the same plugin that extend a base plugin class also written by them. I don't perceive the same can of worms you do. Regardless, say the word and I'll revert that, but if I do, the warning for trying to load `TestPluginGrandChild` (as I called it in the test case I added) before this PR would have said "is not a plugin", and it should be made to say "Must extend Plugin directly"

> would trip the "Class is a plugin but has no plugin descriptor" warning.

It does not trigger that warning. I didn't think it would since `pluginClass.getAnnotation(PluginDescriptor.class);` is based on its runtime class, but I also tested it.

**Comment by Adam-** [comment]
It has to trip it because 

```java
		List<Class<?>> plugins = classPath.getTopLevelClassesRecursive(PLUGIN_PACKAGE).stream()
			.map(ClassInfo::load)
			.collect(Collectors.toList());
```
is all classes in the plugins package, and both classes are in the package.

But regardless, what is it you are trying to do that you think is best solved by having subclassing of plugins?

**Comment by LootBagger** [comment]
Nothing I am doing, nor have plans to do requires extending a class which extends Plugin. I don't need that feature, it seemed like an oversight to not allow it, but I may have been mistaken. I don't know which classes you are referring to when you say "both classes are in the package". Seems like `BasePlugin` and `ChildPlugin` I thought they were examples and do not exist in the Runelite project. And if they did exist, they could exist in a plugin hub plugin, which would not have that problem.

**Comment by LootBagger** [comment]
I made a change so as to load only direct children of `Plugin` in `loadCorePlugins`. Adam, you don't seem sold on this sub-subclassing of Plugin business, so I'm considering modifying this PR to instead issue a more specific diagnostic when trying to load a sub-subclass of plugin, but still not allow/load it. If this PR is scrapped, the change to the `topologicalSort` method is pretty important - iterator invalidation causes a crash if the code is slightly modified, but does not cause a crash as is, obviously.

File: runelite-client/src/main/java/net/runelite/client/plugins/PluginManager.java [hunk 1]
```diff
@@ -648,9 +649,9 @@ private void unschedule(Plugin plugin)
 			T n = it.next();
 			it.remove();
 
-			l.add(n);
+			sorted.add(n);
 
-			for (T m : graphCopy.successors(n))
+			for (T m : new HashSet<>(graphCopy.successors(n)))
```
[the remark this answers is not in the document]
Comment (Adam-) [reply runelite-client/src/main/java/net/runelite/client/plugins/PluginManager.java:654]: I do agree there is probably a problem to fix here. Unfortunately your code to fix it is in a commit with 100+ other lines of unrelated changes which I am less certain of.

[the remark this answers is not in the document]
Comment (LootBagger) [reply runelite-client/src/main/java/net/runelite/client/plugins/PluginManager.java:654]: I'm happy to either split the changes to `topologicalSort` into their own PR or make changes as you see fit to this PR.

[the remark this answers is not in the document]
Comment (Adam-) [reply runelite-client/src/main/java/net/runelite/client/plugins/PluginManager.java:654]: A PR with a commit to fix it + also an explanation of what configuration causes it to break would be most useful

**Comment by LootBagger** [comment]
#14668 fixed the `ConcurrentModificationException` bug in `topologicalSort` and the rest of this PR isn't worth the trouble.

**Pull request closed without merging**